Source/Core/Core/IOS/Network/Socket.cpp

```diff
@@ -730,6 +733,92 @@ void WiiSockMan::Update()
       elem.second.Update(false, false, false);
     }
   }
+  UpdatePollCommands();
+}
+
+void WiiSockMan::UpdatePollCommands()
+{
+  static constexpr int error_event = (POLLHUP | POLLERR);
+
+  if (pending_polls.empty())
+    return;
+
+  const auto now = std::chrono::high_resolution_clock::now();
+  const auto elapsed_d = std::chrono::duration_cast<std::chrono::milliseconds>(now - last_time);
+  const auto elapsed = elapsed_d.count();
+  last_time = now;
+
+  for (auto& pcmd : pending_polls)
+  {
+    // Don't touch negative timeouts
+    if (pcmd.timeout > 0)
+      pcmd.timeout = std::max<s64>(0, pcmd.timeout - elapsed);
+  }
+
+  pending_polls.erase(
+      std::remove_if(
+          pending_polls.begin(), pending_polls.end(),
+          [this](auto& pcmd) {
+            const auto request = Request(pcmd.request_addr);
+            auto& pfds = pcmd.wii_fds;
+            int count = 0;
+            // Happens only on savestate load
+            if (pcmd.wii_fds[0].revents & error_event)
+            {
+              count = static_cast<int>(pcmd.wii_fds.size());
+            }
+            else
+            {
+              // Update FDs to a negative FD in case they were closed:
+              // sockets that were closed during SO_POLL on IOS are ignored
+              for (u32 i = 0; i < pfds.size(); ++i)
```

Comment:
According to the hardware test, you don't need to take care of preserving the original order. All entries in the set will be set properly. So you can use ```std::copy_if``` or ```std::partition``` directly on the vector (without changing the fd to a negative value) and use ```std::distance``` or similar to get the size that will be used by ```poll```.

Reply:
I'm not sure what this means. I need to somehow preserve the order of the pollfd vector to set the correct revents fields when SO_POLL completes, and I'm changing fds to a negative value because I can't pass invalid fds to poll.
I also need to restore the ordering after each poll because SO_POLL may not actually complete (count == 0). 
How do I properly set the revents field when SO_POLL completes if `pfds` doesn't have the same order as what was originally passed to SO_POLL?

Reply:
I thought there might be a simpler solution but since we have to write back `revents` based on the index there might not be. 

However, regarding the negative fds, you can perform the check in `std::partition` rather than assigning a negative fd. Since you move valid fds to the beginning and pass the appropriate size to `poll`, invalid ones aren't passed to poll if I understand the code correctly.